api/services/bulk-docs.js

```diff
@@ -26,45 +19,96 @@ const checkForDuplicates = docs => {
-        return utils.updateParentContacts(docsToDelete);
-      })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+const deleteDocs = (docsToDelete, deletionAttemptCounts, updateAttemptCounts, cumulativeResult = []) => {
+  const finishedModifications = cumulativeResult.map(docUpdate => docUpdate.id);
+  const parentMap = {};
+  let docsToUpdate;
+  let docsToModify;
+  docsToDelete.forEach(doc => doc._deleted = true);
+  incrementAttemptCounts(docsToDelete, deletionAttemptCounts);
+  return utils.updateParentContacts(docsToDelete, parentMap)
+    .then(updatedParents => {
+      docsToUpdate = updatedParents;
+      incrementAttemptCounts(docsToUpdate, updateAttemptCounts);
+      docsToModify = docsToDelete.concat(docsToUpdate).filter(doc => !finishedModifications.includes(doc._id));
+      return db.medic.bulkDocs(docsToModify);
+    })
+    .then(result => {
+      result = result.filter(docUpdate => docUpdate.status !== 404);
+      const deletionFailures = [];
+      const updateFailures = [];
+      const deletionIds = docsToDelete.map(doc => doc._id);
+      const updateIds = docsToUpdate.map(doc => doc._id);
+      const errorIds = result
+        .map((docUpdate, index) => docUpdate.error && (docUpdate.id || docsToModify[index]._id))
+        .filter(id => id);
+      errorIds.forEach(id => {
+        if (deletionIds.includes(id) && deletionAttemptCounts[id] < 4) {
+          deletionFailures.push(id);
+          result = result.filter(docUpdate => docUpdate.id !== id);
+        } else if (updateIds.includes(id) && updateAttemptCounts[id] < 4) {
+          updateFailures.push(id);
+          result = result.filter(docUpdate => docUpdate.id !== id);
+        }
+      });
 
-        res.type('application/json');
-        res.setHeader('X-Content-Type-Options', 'nosniff');
-        res.write('[');
+      cumulativeResult = cumulativeResult.concat(result);
+      if (deletionFailures.length > 0 || updateFailures.length > 0) {
+        return fetchDocs(deletionFailures)
+          .then(docsToDelete => {
+            const updatesToRetry = updateFailures.map(id => parentMap[id]);
+            return deleteDocs(docsToDelete.concat(updatesToRetry), deletionAttemptCounts, updateAttemptCounts, cumulativeResult);
+          });
+      }
+      return cumulativeResult;
+    });
+};
 
-        return batches.reduce((promise, batch, index) => {
-          return promise.then(() => generateBatchPromise(batch, res, { isFinal: index === batches.length - 1 }));
-        }, Promise.resolve([]));
-      })
+module.exports = {
+  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
+    checkForDuplicates(docs);
+    const ids = docs.map(doc => doc._id);
+    const batches = generateBatches(ids, batchSize);
+    const deletionAttemptCounts = {};
+    const updateAttemptCounts = {};
```

Comment:
I don't think you need these two objects at this level?

Reply:
You're right they can go into the reduce below 👍  (but definitely can't be initialised within `deleteDocs`